packages/frontpage/lib/api/post.ts

```diff
@@ -0,0 +1,84 @@
+import "server-only";
+import * as db from "../data/db/post";
+import * as atproto from "../data/atproto/post";
+import { ensureUser, getBlueskyProfile } from "../data/user";
+import { DataLayerError } from "../data/error";
+import { sendDiscordMessage } from "../discord";
+import { invariant } from "../utils";
+import { TID } from "@atproto/common-web";
+
+export type ApiCreatePostInput = {
+  title: string;
+  url: string;
+  createdAt: Date;
+};
+
+export async function createPost({
+  title,
+  url,
+  createdAt,
+}: ApiCreatePostInput) {
+  const user = await ensureUser();
+
+  const rkey = TID.next().toString();
+  try {
+    const dbCreatedPost = await db.createPost({
+      post: { title, url, createdAt },
+      rkey,
+      authorDid: user.did,
+    });
+    invariant(dbCreatedPost, "Failed to insert post in database");
+
+    const { cid } = await atproto.createPost({
+      title: title,
+      url: url,
+      rkey,
+    });
+
+    invariant(cid, "Failed to create comment, rkey/cid missing");
+
+    await db.updatePost({ authorDid: user.did, rkey, cid });
+
+    const bskyProfile = await getBlueskyProfile(user.did);
+
+    await sendDiscordMessage({
+      embeds: [
+        {
+          title: "New post on Frontpage",
+          description: title,
+          url: `https://frontpage.fyi/post/${user.did}/${rkey}`,
+          color: 10181046,
+          author: bskyProfile
+            ? {
+                name: `@${bskyProfile.handle}`,
+                icon_url: bskyProfile.avatar,
+                url: `https://frontpage.fyi/profile/${bskyProfile.handle}`,
+              }
+            : undefined,
+          fields: [
+            {
+              name: "Link",
+              value: url,
+            },
+          ],
+        },
+      ],
+    });
+
+    return { rkey, cid };
+  } catch (e) {
+    await db.deletePost({ authorDid: user.did, rkey });
+    throw new DataLayerError(`Failed to create post: ${e}`);
+  }
+}
+
+export async function deletePost({ rkey }: db.DeletePostInput) {
+  const user = await ensureUser();
+
+  try {
+    await db.deletePost({ authorDid: user.did, rkey });
+    await atproto.deletePost(user.did, rkey);
```

Comment:
Swap these.

Reply:
db needs to come first otherwise there is a chance receive hook processes it before the db has had a chance to write

also if the PDS fails then we erase from the db which is fine but if the db fails we have no real fallback for deleting the record (and it feels like an unnecessary call) on the user's PDS. As it stands it can be swapped you just need a lot of contingencies added. 

Any reason in particular you think it should be swapped?

Reply:
Mentioned in my other comment: https://github.com/likeandscribe/frontpage/pull/210#discussion_r1884103310

> db needs to come first otherwise there is a chance receive hook processes it before the db has had a chance to write

This shouldn't be a problem because deletes should be idempotent.

> if the PDS fails then we erase from the db which is fine

Tricky one because both orders have pros and cons but I don't think I agree. The canonical place for this data is the user's repo, that's the most important place for us to delete the record from.

Reply:
I thought you left this comment for all of them but it seems to just be the deletes, which is fine, I don't think the ordering matters in that case but shouldn't that also mean we should do creates first too? In those cases the comment does apply because creating isn't idempotent and needs handling (I can't remember if I've done that).